### PR TITLE
[TChannel part1] Add Baz tchannel client

### PR DESCRIPTION
**Pull request by ChuntaoLu**
This PR adds Baz client/endpoint thrift files, generated code, and manual Baz tchannel client/server implementation.

This PR is stacked on top of https://github.com/uber/zanzibar/pull/26

r: @Raynos @zhengguiwang

### CODE REVIEW

File: codegen/templates/init_clients.tmpl [hunk 1]
```diff
@@ -9,6 +9,9 @@ import (
 	"{{$pkg}}"
 	{{end}}
 	"github.com/uber/zanzibar/runtime"
+
+	// TODO: (lu) remove this, to be generated
```
Comment (Raynos) [inline codegen/templates/init_clients.tmpl]: This is a neat way of adding hand written code without breaking the code gen :)

File: examples/example-gateway/config/production.json [hunk 2]
```diff
@@ -20,6 +20,9 @@
 	"clients.bar.port": 4001,
 	"clients.bar.ip": "127.0.0.1",
 
+	"clients.SimpleService.port": 4002,
```
Comment (Raynos) [inline examples/example-gateway/config/production.json]: lower case for consistency ?

File: examples/example-gateway/idl/github.com/uber/zanzibar/endpoints/baz/baz.thrift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+exception SimpleErr {
+  1: required string message
+}
+
+exception NewErr {
+  1: required string message
+}
+
+service SimpleService {
```
Comment (Raynos) [inline examples/example-gateway/idl/github.com/uber/zanzibar/endpoints/baz/baz.thrift]: This endpoint will need http annotations for it to work ?

File: runtime/tchannel_server.go [hunk 4]
```diff
@@ -27,15 +27,9 @@ import (
 	"github.com/pkg/errors"
 	"github.com/uber-go/zap"
 	"github.com/uber/tchannel-go"
-)
```
Comment (Raynos) [inline runtime/tchannel_server.go]: Should we move this file into the tchannel folder ?

File: examples/example-gateway/clients/baz/baz.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
+// Code generated by zanzibar
+// @generated
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: If this file is not generated then dont add the generated comments ;)

File: examples/example-gateway/clients/baz/baz.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+// NewClient returns a new http client for service Bar.
+func NewClient(config *zanzibar.StaticConfig, gateway *zanzibar.Gateway) *BazClient {
+	// TODO: (lu) get clientName from config
+	// clientName := config.MustGetString("clients.baz.clientName")
+	clientName := "BazClient"
+
+	// TODO: (lu) client channel opts
+	ch, err := tchannel.NewChannel(clientName, nil)
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: We do not want to allocate new channels here.

There needs to be a `gateway.Channel` and we should allocate a sub channel here.

File: examples/example-gateway/clients/baz/baz.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	if err != nil {
+		gateway.Logger.Error("Failed to create tchannel",
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: If creating a client can fail then we need to refactor our client interfaces.

We should have something like

```
NewX(config, gateway) *XClient
(*XClient) Bootstrap() error
(*XClient) Close()
```

Aka, if a client has some bootstrapping logic that can fail we need an explicit bootstrap method ( or init method, methodname TBD. )

File: examples/example-gateway/clients/baz/baz.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+			zap.String("error", err.Error()))
+	}
+
+	ip := config.MustGetString("clients.SimpleService.ip")
+	port := config.MustGetInt("clients.SimpleService.port")
+	ch.Peers().Add(ip + ":" + strconv.Itoa(int(port)))
+
+	// TODO: (lu) get serviceName from config
+	// serviceName := config.MustGetString("clients.baz.serviceName")
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: ServiceName is a thrift detail so its known from the thrift file and/or client-config.json.

File: examples/example-gateway/clients/baz/baz.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+	return &BazClient{
+		SimpleServiceClient{
+			thriftService: serviceName,
+			client:        client,
+		},
+	}
+}
+
+// SimpleServiceClient is the client to talk to SimpleService backend.
+type SimpleServiceClient struct {
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: Can we merge this client with `BazClient` ?

File: examples/example-gateway/clients/baz/baz.go [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+		default:
+			err = errors.New("received no result or unknown exception for SimpleFuture")
+		}
+	}
+
+	return respHeaders, err
+}
+
+// SimpleServiceServer is the Baz backend service
+type SimpleServiceServer struct {
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: Who uses `SimpleServiceServer` ? It looks like this code doesn't belong here.

Can we move it to `endpoints/baz/baz.go` or something ?

File: examples/example-gateway/clients/baz/baz.go (hunk 8 above)
Comment (ChuntaoLu) [inline examples/example-gateway/clients/baz/baz.go]: yeah, I removed the comment in the following PR https://github.com/uber/zanzibar/pull/30/files#diff-50f740ccde9b3f6a874ee45ca46b1613L53

File: examples/example-gateway/clients/baz/baz.go (hunk 6 above)
Comment (ChuntaoLu) [inline examples/example-gateway/clients/baz/baz.go]: make sense, will make changes accordingly

File: examples/example-gateway/clients/baz/baz.go (hunk 7 above)
Comment (ChuntaoLu) [inline examples/example-gateway/clients/baz/baz.go]: if we pass in a top level channel via gateway and not allocate new channel  here, we can probably don't have failure cases.

File: examples/example-gateway/clients/baz/baz.go (hunk 9 above)
Comment (ChuntaoLu) [inline examples/example-gateway/clients/baz/baz.go]: Yes

File: examples/example-gateway/clients/baz/baz.go (hunk 10 above)
Comment (ChuntaoLu) [inline examples/example-gateway/clients/baz/baz.go]: the test backend use it to create mock backends. Yeah, it does not belong here, but I don't think `endpoints/baz/baz.go` is a good fit either. I should have put a todo here.

File: examples/example-gateway/config/production.json (hunk 2 above)
Comment (ChuntaoLu) [inline examples/example-gateway/config/production.json]: I imagine we will pull out the service name from thrift or client config, and use it as key to get from the above config, https://github.com/uber/zanzibar/pull/27/files#diff-50f740ccde9b3f6a874ee45ca46b1613R49. We can follow convention here, but a conversion is needed in the middle somewhere, which I am not in favor of.

And the `clients` prefix in `clients.xxxx.port` does not really make sense, it's clearly downstream service configs. In this sense, it's arguably more consistent as is the same as the service name defined in thrift.

File: examples/example-gateway/idl/github.com/uber/zanzibar/endpoints/baz/baz.thrift (hunk 3 above)
Comment (ChuntaoLu) [inline examples/example-gateway/idl/github.com/uber/zanzibar/endpoints/baz/baz.thrift]: Right, I am deferring it a bit

File: runtime/tchannel_server.go (hunk 4 above)
Comment (ChuntaoLu) [inline runtime/tchannel_server.go]: Yeah, we can even merge it into `runtime/tchannel/server.go`, which is probably even better?

File: examples/example-gateway/config/production.json (hunk 2 above)
Comment (Raynos) [inline examples/example-gateway/config/production.json]: We need to use something from client-config.json, like the "name" field ( https://github.com/uber/zanzibar/blob/master/examples/example-gateway/clients/googlenow/client-config.json#L2 ).

The thrift file is not the source of truth, the client-config is.

File: runtime/tchannel_server.go (hunk 4 above)
Comment (Raynos) [inline runtime/tchannel_server.go]: cool, sounds good.

File: examples/example-gateway/clients/baz/baz.go (hunk 10 above)
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: cool.

File: runtime/tchannel_server.go (hunk 4 above)
Comment (ChuntaoLu) [inline runtime/tchannel_server.go]: I will make the merge when updating part 3 PR.

File: examples/example-gateway/clients/baz/baz.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+// NewClient returns a new http client for service Bar.
+func NewClient(config *zanzibar.StaticConfig, gateway *zanzibar.Gateway) *BazClient {
+	// TODO: (lu) get serviceName from thrift
+	serviceName := "SimpleService"
+
+	ch := gateway.Channel
+
+	ip := config.MustGetString("clients.baz.ip")
+	port := config.MustGetInt("clients.baz.port")
+	ch.Peers().Add(ip + ":" + strconv.Itoa(int(port)))
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: We cannot mutate the `Peers()` of the top channel.

We need to allocate a sub channel.

File: runtime/gateway.go [hunk 12]
```diff
@@ -91,12 +94,19 @@ func CreateGateway(
 		logWriter = opts.LogWriter
 	}
 
+	// TODO: (lu) channel name and opts
+	ch, err := tchannel.NewChannel("Gateway", nil)
```
Comment (Raynos) [inline runtime/gateway.go]: This needs to be allocated in `setupChannel()` not here.

File: runtime/gateway.go [hunk 13]
```diff
@@ -353,11 +356,19 @@ func (gateway *Gateway) setupHTTPServer() error {
 }
 
 func (gateway *Gateway) setupTChannel(config *StaticConfig) error {
+	// TODO: (lu) channel name and opts
+	ch, err := tchannel.NewChannel("Gateway", nil)
```
Comment (Raynos) [inline runtime/gateway.go]: That's not really what I meant. This is duplicate with `runtime/tchannel_server.go:56`

we should use `tchannelServer.Channel`.

Never allocate two channels.

File: examples/example-gateway/clients/baz/baz.go [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+	// TODO: (lu) get serviceName from thrift
+	serviceName := "SimpleService"
+
+	sc := gateway.Channel.GetSubChannel(serviceName)
+
+	ip := config.MustGetString("clients.baz.ip")
+	port := config.MustGetInt("clients.baz.port")
+	sc.Peers().Add(ip + ":" + strconv.Itoa(int(port)))
+
+	client := zt.NewClient(gateway.Channel, serviceName)
```
Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: Got to pass the sub channel in here otherwise it wont work.

Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: Apparently we have to pass both in because `ch.RunWithRetry` ...

Comment (ChuntaoLu) [inline examples/example-gateway/clients/baz/baz.go]: the allocated sub channel is acquired by calling `channel.GetSubSchannel(serviceName)` in `zt.NewClient` call

Comment (Raynos) [inline examples/example-gateway/clients/baz/baz.go]: I guess that's good enough.

File: runtime/tchannel_server.go [hunk 15]
```diff
@@ -72,8 +68,7 @@ func NewTChannelServer(opts *TChannelServerOptions) (*TChannelServer, error) {
 			err)
 	}
 
-	return &TChannelServer{
-		Channel: channel,
-		Logger:  opts.Logger,
-	}, nil
+	gateway.Channel = channel
```
Comment (Raynos) [inline runtime/tchannel_server.go]: mutating the gateway here is naughty, can we move this into `setupChannel()`.

Keeping mutation of the gateway to gateway.go makes debugging easier.

Comment (Raynos) [inline runtime/tchannel_server.go]: Sounds good.

**Pull request merged**